File: baml_language/sdk_tests/crates/python_pydantic2/type_shapes/customizable/roundtrip_tests/test_aliases.py
"""Roundtrip coverage for `baml_sdk.aliases` — type aliases (incl. recursive)."""

import baml_sdk  # noqa: F401  — initializes the BAML runtime
from baml_sdk.aliases import (
    AliasContainer,
    round_trip_string_list,
    round_trip_rec_list,
    round_trip_alias_container,
)


def test_aliases_round_trip_string_list():
    assert round_trip_string_list(s=["a", "b"]) == ["a", "b"]


def test_aliases_round_trip_rec_list():
    # RecList = int | RecList[]
    assert round_trip_rec_list(r=1) == 1
    assert round_trip_rec_list(r=[1, [2, 3]]) == [1, [2, 3]]


def test_aliases_round_trip_alias_container():
    c = AliasContainer(list_field=["x"], rec_field=[1, [2]])
    assert round_trip_alias_container(c=c) == c

File: baml_language/sdk_tests/crates/rust/type_shapes/customizable/roundtrip_tests/test_aliases.rs
//! Roundtrip coverage for `baml_sdk::aliases` — type aliases (incl. recursive).

use baml_sdk::aliases::round_trip_string_list;

#[test]
fn test_aliases_round_trip_string_list() {
    assert_eq!(
        round_trip_string_list(vec!["a".to_string(), "b".to_string()]).unwrap(),
        vec!["a".to_string(), "b".to_string()]
    );
}

#[test]
fn test_aliases_round_trip_rec_list() {
    // RecList = int | RecList[]
    //
    // DIVERGENCE(rust): `RecList` is a recursive *union* alias — unions have
    // no generated Rust representation yet, and recursive aliases are
    // skipped by codegen initially, so this round trip is inexpressible.
    // Intended body once both exist (`RecList` as the generated recursive
    // union):
    //
    //     assert_eq!(round_trip_rec_list(RecList::Int(1)).unwrap(), RecList::Int(1));
    //     let r = RecList::List(vec![
    //         RecList::Int(1),
    //         RecList::List(vec![RecList::Int(2), RecList::Int(3)]),
    //     ]);
    //     assert_eq!(round_trip_rec_list(r.clone()).unwrap(), r);
}

#[test]
fn test_aliases_round_trip_alias_container() {
    // DIVERGENCE(rust): `AliasContainer.rec_field` is the recursive union
    // alias `RecList` (see above), so the struct cannot be constructed yet.
    // Intended body once unions and recursive aliases exist:
    //
    //     let c = AliasContainer {
    //         list_field: vec!["x".to_string()],
    //         rec_field: RecList::List(vec![
    //             RecList::Int(1),
    //             RecList::List(vec![RecList::Int(2)]),
    //         ]),
    //     };
    //     assert_eq!(round_trip_alias_container(c.clone()).unwrap(), c);
}
